docs/architecture/adr-0023-text-array-object-acrs.md

```diff
@@ -0,0 +1,166 @@
+
+### Serialization and Deserialization
+
+Doobie supports mappings between single-dimensional `text[]` and `List[String]`. To construct `"<subject type>;<optional subject id>;<action type>"` strings, we will create a data model to transform a list of access control rules to strings conforming to this structure before inserting to database, which can be easily tested using property based tests.
+
+Text array ACRs should not be returned on any listing endpoints for first class objects. Only object owners are able to access ACRs through `<fist class object>/<object id>/permissions` endpoints, which we have been using, and which will need some adaptions.
+
+### ACR-related actions
+
+For `CRUD` of ACRs on an object, no special handling, other than importing `doobie.postgres._` and `doobie.postgres.implicits._`, will be required since doobie does the mappings. Note that, currently, when deleting a single permission, we update permissions with the rest of other permissions. So this is not a "delete" so to speak but more of a re-assigning. When deleting permission of an object, we will just pass an empty `text[]` to the `acrs` field.
+
+In terms of querying for objects with filters on object level ACRs, postgres array functions such as `array_cat()` and `concat_ws()`, and operators such as `&&` (which determines overlay of two arrays), etc., could potentially be handy.
+
+When listing actions or permissions by subject type and id, which will potentially require going into each string of the text array, we could use postgres array functions such as `unnest()` combined with `like` to match string patterns.
+
+### Migration
+
+Migrations could be a bit tricky and probably could take some time to run. Adding a new field and updating this field in all first-class objects' tables would likely require a novel approach of creating a temp table with the new column, writing to a new permanent table, and dropping the previous table. Further research on it will be in a future issue card.
+
+Note that the `acrs` field will have an empty text array as the default value.
+
+### Examples for datasources
+
+This section shows some initial attempts for adapting to the ACR text array solution. These are for initial tests only and don't represent final implementation details. Datasources could be a simple case, but such adaptations can be applied to other first-class objects in the similar way.
+
+#### Data models
+
+```scala
+@JsonCodec
+case class Datasource(
+  id: UUID,
+  createdAt: java.sql.Timestamp,
+  createdBy: String,
+  modifiedAt: java.sql.Timestamp,
+  modifiedBy: String,
+  owner: String,
+  name: String,
+  visibility: Visibility,
+  composites: Json,
+  extras: Json,
+  bands: Json,
+  licenseName: Option[String],
+  acrs: List[String]
+)
+
+@JsonCodec
+case class ObjectAccessControlRuleCreate (
+  subjectType: SubjectType,
+  subjectIdO: Option[String],
```

Comment:
I think it is better to spell out `Option` in this case, it's really easy to mistake this for just `subjectId` or think this is a typo otherwise.

Reply:
I've been putting `-O` and `-F` and `-IO` on things to indicate containers for a bit -- it's consistent 
doobie's Fragment indicators and I think I've seen it in other typelevel projects as well, but would
have to go digging.

Reply:
Ah ok - if it's standard practice in this area I will just get with the times then :(